aws/net-lb/main.tf

```diff
@@ -0,0 +1,167 @@
+data "aws_subnets" "private" {
+  filter {
+    name   = "vpc-id"
+    values = [var.vpc_id]
+  }
+
+  tags = {
+    Name = "*private*"
+  }
+}
+
+## Get Public subnets from VPC
+data "aws_subnets" "public" {
+  filter {
+    name   = "vpc-id"
+    values = [var.vpc_id]
+  }
+
+  tags = {
+    Name = "*public*"
+  }
+}
+
+# Network Load Balancer 
+
+resource "aws_lb" "network" {
+  name                             = var.nlb_name
+  internal                         = var.nlb_type_internal
+  load_balancer_type               = "network"
+  subnets                          = var.nlb_type_internal ? data.aws_subnets.private.ids : data.aws_subnets.public.ids
+  enable_cross_zone_load_balancing = var.enable_cross_zone_load_balancing
+
+  dynamic "access_logs" {
+    for_each = var.enable_access_logging ? { enabled = true } : {}
+    content {
+      bucket  = var.s3_bucket_id != "" ? var.s3_bucket_id : aws_s3_bucket.fg_nlb_access_logs[0].id
+      enabled = var.enable_access_logging
+    }
+  }
+
+  tags = var.tags
+}
+
+## Optional S3 Bucket Resources for Access Logs
+
+resource "aws_s3_bucket" "fg_nlb_access_logs" {
+  count         = (var.enable_access_logging && (var.s3_bucket_id == "")) ? 1 : 0
+  bucket        = "${var.nlb_name}-access-logs"
+  force_destroy = var.force_destroy_nlb_access_logs
+}
+
+resource "aws_s3_bucket_acl" "fg_nlb_access_logs" {
+  count  = (var.enable_access_logging && (var.s3_bucket_id == "")) ? 1 : 0
+  bucket = aws_s3_bucket.fg_nlb_access_logs[0].id
+  acl    = "private"
+}
+
+resource "aws_s3_bucket_policy" "fg_nlb_access_logs" {
+  count  = (var.enable_access_logging && (var.s3_bucket_id == "")) ? 1 : 0
+  bucket = aws_s3_bucket.fg_nlb_access_logs[0].id
+  policy = data.aws_iam_policy_document.allow_nlb_write_to_bucket[0].json
+}
+
+data "aws_elb_service_account" "main" {}
+
+data "aws_iam_policy_document" "allow_nlb_write_to_bucket" {
+  count = (var.enable_access_logging && (var.s3_bucket_id == "")) ? 1 : 0
+
+  version = "2012-10-17"
+  statement {
+    principals {
+      type        = "AWS"
+      identifiers = [data.aws_elb_service_account.main.arn]
+    }
+    effect = "Allow"
+    actions = [
+      "s3:PutObject",
+    ]
+    resources = [
+      "${aws_s3_bucket.fg_nlb_access_logs[0].arn}/*"
+    ]
+  }
+}
+
+## Default Target Group
+resource "aws_lb_target_group" "default" {
+  name        = "${var.nlb_name}-tg"
+  port        = 80
+  protocol    = var.default_target_group_protocol
+  vpc_id      = var.vpc_id
+  target_type = var.default_target_type
+
+  health_check {
+    timeout             = var.health_check.timeout
+    interval            = var.health_check.interval
+    port                = var.health_check.port
+    protocol            = var.health_check.protocol
+    unhealthy_threshold = var.health_check.unhealthy_threshold
+    healthy_threshold   = var.health_check.healthy_threshold
+  }
+
+  tags = var.tags
+}
+
+## Listeners
+resource "aws_lb_listener" "http" {
```

Comment:
Might it make sense to only deploy a HTTP listener if it's an internal LB? I say this because then we're _always_ using TLS by default when using this as an external LB.

An extra supercalifragilisticexpialidocious solution would be to actually forward to the HTTPS listener if this is an external LB.

Reply:
So I think currently we are following a similar flow - if the NLB is internal, the user can choose whether or not to use TLS, and an HTTPS listener will only be deployed if they want TLS. If the NLB is external, we are enforcing TLS by default.

Will move conditions out into locals blocks in a separate PR to improve readability.